lls/main.c

```diff
@@ -285,6 +291,130 @@ match_nd(struct rte_mbuf *pkt, struct gatekeeper_if *iface)
 	return 0;
 }
 
+static void
+xmit_icmpv6_reply(struct gatekeeper_if *iface, struct rte_mbuf *pkt)
+{
+	struct ether_addr eth_addr_tmp;
+	struct ether_hdr *icmp_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	uint8_t ip6_addr_tmp[16];
+	struct ipv6_hdr *icmp_ipv6;
+	struct icmpv6_hdr *icmpv6_hdr;
+
+	ether_addr_copy(&icmp_eth->s_addr, &eth_addr_tmp);
+	ether_addr_copy(&icmp_eth->d_addr, &icmp_eth->s_addr);
+	ether_addr_copy(&eth_addr_tmp, &icmp_eth->d_addr);
+	if (iface->vlan_insert)
+		fill_vlan_hdr(icmp_eth, iface->vlan_tag_be, ETHER_TYPE_IPv6);
+
+
+	/* Set-up IPv6 header. */
+	icmp_ipv6 = (struct ipv6_hdr *)pkt_out_skip_l2(iface, icmp_eth);
+
+	/*
+	 * The IP Hop Limit field must be 255 as required by
+	 * RFC 4861, sections 7.1.1 and 7.1.2.
+	 */
+	icmp_ipv6->hop_limits = 255;
+	rte_memcpy(ip6_addr_tmp, icmp_ipv6->src_addr, sizeof(ip6_addr_tmp));
+	rte_memcpy(icmp_ipv6->src_addr, icmp_ipv6->dst_addr,
+		sizeof(icmp_ipv6->src_addr));
+	rte_memcpy(icmp_ipv6->dst_addr, ip6_addr_tmp,
+		sizeof(icmp_ipv6->dst_addr));
+
+	/* Set-up ICMPv6 header. */
+	icmpv6_hdr = (struct icmpv6_hdr *)&icmp_ipv6[1];
+	icmpv6_hdr->type = ICMPV6_ECHO_REPLY_TYPE;
+	icmpv6_hdr->code = ICMPV6_ECHO_REPLY_CODE;
+	icmpv6_hdr->cksum = 0; /* Calculated below. */
+
+	icmpv6_hdr->cksum = rte_ipv6_icmpv6_cksum(icmp_ipv6, icmpv6_hdr);
+}
+
+static int
+reply_ping6(struct rte_mbuf **pkts, unsigned int num_pkts,
+	struct gatekeeper_if *iface)
+{
+	unsigned int i;
+	uint16_t tx_queue = iface == &lls_conf.net->front ?
+		lls_conf.tx_queue_front : lls_conf.tx_queue_back;
+	uint16_t num_pkts_sent = 0;
+	struct rte_mbuf *reply_bufs[num_pkts];
+	struct token_bucket_ratelimit_state *rs =
+		iface == &lls_conf.net->front ? &lls_conf.front_icmp_rs :
+		&lls_conf.back_icmp_rs;
+
+	RTE_VERIFY(num_pkts <= lls_conf.mailbox_max_pkt_sub);
+
+	for (i = 0; i < num_pkts; i++) {
+		if (tb_ratelimit_allow(rs)) {
```

Comment:
For this case, it's worth adding `tb_ratelimit_allow_n()` to `lib/ratelimit.c`; `tb_ratelimit_allow_n()` would return the number of granted packets. And `tb_ratelimit_allow()` would become an inline function calling `tb_ratelimit_allow_n()` passing 1.